README.md
# psd-guides

[![NPM Version][npm-image]][npm-url]
[![Minzipped size][bundlephobia-image]][bundlephobic-url]
[![License][license-image]][license-url]
![Test CI][github-actions-test-image]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![Dependencies][david-image]][david-url]
[![Dev Dependencies][david-dev-image]][david-dev-url]
[![Total alerts][lgtm-image]][lgtm-url]
[![Language grade: JavaScript][lgtm-grade-image]][lgtm-grade-url]

http://noeldelgado.github.io/psd-guides/

psd-guides is a simple script to draw photoshop-like guides.

Can be useful during slicing phase to accomplish pixel-perfect web layouts.

## Installation
**NPM**

	npm install psd-guides --save

## Usage
```javascript
new PSDGuides({
  canvasWidth : 1000,
  horizontalGuides : [20, "355 * 2", 250, 20],
  vericalGuides : [50, "100 * 2", "250 * 2", "50 * 3"]
}).activate();
```

**Tip:** While defining your guides, if you have similar values that needs to be repeated several times, instead of writing them all you can use the `*` character followed by the number you want it to be repeated, for instance: `vericalGuides : [10, 10, 10, 10, 10, 20, 30, 20, 30, 20, 30]` can be written as `verticalGuides : ["10 * 5", "(20, 30) * 3"]`

## Defaults
```javascript
{
  canvas : document.body,                 // [DOMElement]
  canvasWidth : 0,                        // [Integer] (pixels)
  alignment : "center",                   // [String] "center"|"left"|"right"
  backColor : "rgba(132, 170, 234, .25)", // [String] Any valid color format
  lineColor : "rgba(73, 141, 255, 1)",    // [String] Any valid color format
  horizontalGuides : [],                  // [Array]
  verticalGuides : [],                    // [Array]
  zindex : 9999                           // [Number]
}
```

## API

### activate
```javascript
/**
 * Display the guides.
 * @property activate <public> [Function]
 * @return this [PSDGuides]
 */

var psd = new PSDGuides({
	canvasWidth : 1000,
	horizontalGuides : [20, "355 * 2", 250],
	verticalGuides : [50, "100 * 2", "250 * 2", "50 * 3"]
});

psd.activate();
```

### deactivate
```javascript
/**
 * Hide the guides.
 * @property deactivate <public> [Function]
 * @return this [PSDGuides]
 */

psd.deactivate();
```

### update
```javascript
/**
 * Update the width and height of psd-guides container,
 * remove and create the guides using the guides array references.
 * @property update <public> [Function]
 * @return this [PSDGuides]
 */

psd.update();
```

### destroy
```javascript
/**
 * Clean all references to other objects and remove DOMElements.
 * @property destroy <public> [Function]
 * @return null
 */

psd.destroy();
// => null
```

### # Removing guides
When guides are removed, you need to call the `update` or `activate` method to reflect the changes.

### removeHorizontalGuides
```javascript
/**
 * Clear the horizontal guides array reference.
 * @property removeHorizontalGuides <public> [Function]
 * @return this [PSDGuides]
 */

psd.removeHorizontalGuides();
// console.log( psd.getHorizontalGuides() );
// => []
// console.log( psd.getVerticalGuides() );
// => [50, 100, 100, 250, 250, 50, 50, 50]
```

### removeVerticalGuides
```javascript
/**
 * Clear the vertical guides array reference.
 * @property removeVerticalGuides <public> [Function]
 * @return this [PSDGuides]
 */

psd.removeVerticalGuides();
// console.log( psd.getVerticalGuides() );
// => []
```

### removeGuides
```javascript
/**
 * Clear both horizontal and vertical array references.
 * @property removeGuides <public> [Function]
 * @return this [PSDGuides]
 */

psd.removeGuides().update();
```

### # Adding guides
When adding guides, you need to call the `update` or `activate` method after to reflect the changes.

### addHorizontalGuides
```javascript
/**
 * Add guides to the _horizontalGuides Array holder.
 * @property addHorizontalGuides <public> [Function]
 * @argument guides <required> [Array]
 * @return this [PSDGuides]
 */

psd.addHorizontalGuides([20, "355 * 2", 250]).update();
// console.log( psd.getHorizontalGuides() );
// => [20, 355, 355, 250]
```

### addVericalGuides
```javascript
/**
 * Add guides to the _verticalGuides Array holder.
 * @property addVerticalGuides <public> [Function]
 * @argument guides <required> [Array]
 * @return this [PSDGuides]
 */

psd.addVerticalGuides([50, "100 * 2", "250 * 2", "50 * 3"]).update();
// console.log( psd.getVerticalGuides() );
// => [50, 100, 100, 250, 250, 50, 50, 50]
```

### # Getting the guides
Get current saved guides.

### getHorizontalGuides
```javascript
/**
 * Return the current saved horizontal guides.
 * @property getHorizontalGuides <public> [Function]
 * @return this._horizontalGuides [Array]
 */

psd.getHorizontalGuides();
// => [20, 355, 355, 250]
```

### getVerticalGuides
```javascript
/**
 * Return the current saved vertical guides.
 * @property getVerticalGuides <public> [Function]
 * @return this._verticalGuides [Array]
 */

psd.getVerticalGuides();
// => [50, 100, 100, 250, 250, 50, 50, 50]
```

## Examples (960 Grid System)

Applying 960 grid system

### 12-column grid
```javascript
new PSDGuides({
    canvasWidth : 960,
    horizontalGuides : ["(10, 60, 10) * 12"]
}).activate()
```

### 16-column grid
```javascript
new PSDGuides({
    canvasWidth : 960,
    horizontalGuides : ["(10, 40, 10) * 16"]
}).activate();
```

### 24-column grid
```javascript
new PSDGuides({
    canvasWidth : 960,
    horizontalGuides : ["(10, 20, 10) * 24"]
}).activate()
```

There are more examples inside the `examples` folder.

[npm-image]: https://img.shields.io/npm/v/psd-guides.svg?logo=npm&label=NPM
[npm-url]: https://www.npmjs.com/package/psd-guides
[license-image]: https://img.shields.io/npm/l/psd-guides.svg
[license-url]: https://github.com/noeldelgado/psd-guides/blob/master/LICENSE
[github-actions-test-image]: https://github.com/noeldelgado/psd-guides/workflows/Test%20CI/badge.svg
[bundlephobia-image]: https://img.shields.io/bundlephobia/minzip/psd-guides?label=size
[bundlephobic-url]: https://bundlephobia.com/result?p=psd-guides
[snyk-image]: https://snyk.io/test/npm/psd-guides/badge.svg
[snyk-url]: https://snyk.io/test/npm/psd-guides
[david-image]: https://img.shields.io/david/noeldelgado/psd-guides.svg
[david-url]: https://david-dm.org/noeldelgado/psd-guides
[david-dev-image]: https://img.shields.io/david/dev/noeldelgado/psd-guides.svg
[david-dev-url]: https://david-dm.org/noeldelgado/psd-guides?type=dev
[lgtm-image]: https://img.shields.io/lgtm/alerts/g/noeldelgado/psd-guides.svg?logo=lgtm&logoWidth=18
[lgtm-url]: https://lgtm.com/projects/g/noeldelgado/psd-guides/alerts/
[lgtm-grade-image]: https://img.shields.io/lgtm/grade/javascript/g/noeldelgado/psd-guides.svg?logo=lgtm&logoWidth=18
[lgtm-grade-url]: https://lgtm.com/projects/g/noeldelgado/psd-guides/context:javascript


CHANGELOG.md
# Changelog

## [Unreleased]

(feature/tb) [#5](https://github.com/noeldelgado/psd-guides/pull/5) - 2020-06-03
### Added
- test-ci workflow
- badges
  - bundlephobia
  - license
  - Github Action test-ci
  - snyk
  - david
  - lgtm

(feature/t) [#4](https://github.com/noeldelgado/psd-guides/pull/4) - 2020-06-03
### Added
- .editorconfig [bd2c6ee](https://github.com/noeldelgado/psd-guides/commit/bd2c6eeca7141092c3aedcada29bb745e6c4bdcf)
- tests [bc32c90](https://github.com/noeldelgado/psd-guides/commit/bc32c90e41b6b5527d629d7c72aa9828bbade25e) [ede4a6e](https://github.com/noeldelgado/psd-guides/commit/ede4a6e5c4f5d9aec517a7993d2969a817f0be46) [3cc4797](https://github.com/noeldelgado/psd-guides/commit/3cc47973472edcaa9203778cc20af5fcb0561a30)

(feature/rbwp) [#3](https://github.com/noeldelgado/psd-guides/pull/3) - 2020-06-03
### Removed
- bower distribution

(feature/npmfiles) [#2](https://github.com/noeldelgado/psd-guides/pull/2) - 2020-06-03
### Fixed
- distributed NPM files

(feauture/li) [#1](https://github.com/noeldelgado/psd-guides/pull/1) - 2020-06-03
### Added
- license
- changelog

## [1.0.1]
### Fix
- commonjs support
- demos

### Removed
- build step

## [1.0.0]
- initial release

[0.1.1]: https://github.com/noeldelgado/psd-guides/compare/v1.0.0...v1.0.1
[0.1.0]: https://github.com/noeldelgado/psd-guides/releases/tag/v1.0.0


bundle.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
/**
 * psd-guides - Draw photoshop-like guides
 * @version v1.0.1
 * @link https://github.com/noeldelgado/psd-guides
 * @license MIT
 */
(function () {
  function PSDGuides(config) {
    this.config = {
      canvas: document.body,
      canvasWidth: 0,
      alignment: "center",
      backColor: "rgba(132, 170, 234, 0.25)",
      lineColor: "rgba(73, 141, 255, 1)",
      horizontalGuides: [],
      verticalGuides: [],
      zindex: 9999
    };
    this._ui = null;
    this._verticalGuides = null;
    this._horizontalGuides = null;
    this.__resizeHandler = null;
    this._reGroup = null;
    this._reSplit = null;
    this.active = null;

    Object.keys(config).forEach(function(propertyName) {
      this.config[propertyName] = config[propertyName];
    }, this);

    this._init();
  }

  PSDGuides.prototype._init = function _init() {
    this._ui = {};
    this.active = false;
    this._verticalGuides = [];
    this._horizontalGuides = [];
    this._reGroup = new RegExp("[()\\s]","g");
    this._reSplit = new RegExp("[,*]","g");

    this._ui.documentElement = window.document.documentElement;
    this._ui.body = window.document.body;
    this._ui.wrapper = document.createElement("div");
    this._ui.hContainer = document.createElement("div");
    this._ui.vContainer = document.createElement("div");

    this._setStyles(this._ui.wrapper, {
      display: "none",
      position: "absolute",
      top: 0,
      zIndex: this.config.zIndex,
      backgroundColor: this.config.backColor
    })._setStyles(this._ui.hContainer, {
      position: "absolute",
      top: 0,
      height: "100%",
      borderLeft: "1px dotted " + this.config.lineColor,
      borderRight: "1px dotted " + this.config.lineColor
    })._setStyles(this._ui.vContainer, {
      position: "absolute",
      top: 0,
      height: "100%",
      width: "100%"
    });

    this._ui.wrapper.className = "psd-guides-wrapper";
    this.config.canvas.appendChild(this._ui.wrapper);

    this._bindEvents();
    this._overrideCSS();
    this.addVerticalGuides(this.config.verticalGuides);
    this.addHorizontalGuides(this.config.horizontalGuides);
    this.update();

    return this;
  };

  PSDGuides.prototype._bindEvents = function _bindEvents() {
    this.__resizeHandler = this._resizeHandler.bind(this);
    window.addEventListener("resize", this.__resizeHandler, false);

    return this;
  };

  PSDGuides.prototype._resizeHandler = function _resizeHandler() {
    if (this.active) {
      this._hide();
      window.clearTimeout(this._resizeTimer);
      this._resizeTimer = window.setTimeout(function() {
        this.update().activate();
      }.bind(this), 250);
    }
  };

  /**
   * Add guides to the _verticalGuides Array holder.
   * @property addVerticalGuides <public> [Function]
   * @argument guides <required> [Array]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype.addVerticalGuides = function addVerticalGuides(guides) {
    this._verticalGuides = this._verticalGuides.concat(this._getParsedGuides(guides));

    return this;
  };

  /**
   * Add guides to the _horizontalGuides Array holder.
   * @property addHorizontalGuides <public> [Function]
   * @argument guides <required> [Array]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype.addHorizontalGuides = function addHorizontalGuides(guides) {
    this._horizontalGuides = this._horizontalGuides.concat(this._getParsedGuides(guides));

    return this;
  };

  /**
   * Check if the guides needs to be translated into Numbers.
   * parsed = this._getParsedGuides([100, "200 * 4"], 100);
   * => [100, 200, 200, 200, 200, 100]
   * @property _getParsedGuides <private> [Function]
   * @argument guides <required> [Array]
   * @return _parsedGuides [Array]
   */
  PSDGuides.prototype._getParsedGuides = function _getParsedGuides(guides) {
    var _parsedGuides = [];

    guides.map(function (guide) {
      if ((typeof guide === "string") && (guide.indexOf("*") !== -1)) {
        var values, times, length, i, j;

        values = guide.replace(this._reGroup, "").split(this._reSplit);
        times = values.pop();
        length = values.length;

        for (i = 0; i < times; i++) {
          for (j = 0; j < length; j++) {
            _parsedGuides.push(~~values[j]);
          }
        }

        return;
      }

      _parsedGuides.push(~~guide);
    }, this);

    return _parsedGuides;
  };

  /**
   * Create the vertical guides DOMElements and append them to its parent.
   * @property _createVerticalLines <private> [Function]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype._createVerticalLines = function _createVerticalLines() {
    var fragment, newDocHeight;

    fragment = document.createDocumentFragment();
    newDocHeight = 0;

    this.getVerticalGuides().map(function (guide) {
      this._appendLine(fragment, {
        height: guide + "px",
        borderBottom: "1px dotted " + this.config.lineColor
      });

      newDocHeight += guide;
    }, this);

    newDocHeight = Math.max(this._ui.body.clientHeight, newDocHeight);

    this._ui.wrapper.style.height = newDocHeight;
    this._ui.vContainer.appendChild(fragment);
    this._ui.wrapper.appendChild(this._ui.vContainer);

    return this;
  };

  /**
   * Create the horizontal guides DOMElements and append them to its parent.
   * @prototype _createHorizontalLines <private> [Object]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype._createHorizontalLines = function _createHorizontalLines() {
    var fragment, siteWidth, coveredWidth;

    fragment = document.createDocumentFragment();
    siteWidth = this.config.canvasWidth;
    coveredWidth = 0;

    this._setStyles(this._ui.hContainer, {
      width: (siteWidth) + "px",
      marginLeft: this._getAligment(siteWidth) + "px"
    }).getHorizontalGuides().map(function (guide) {
      this._appendLine(fragment, {
        position: "absolute",
        left: coveredWidth + "px",
        width: guide + "px",
        height: "100%",
        borderRight: "1px dotted " + this.config.lineColor
      });

      coveredWidth += guide;
    }, this);

    this._ui.hContainer.appendChild(fragment);
    this._ui.wrapper.appendChild(this._ui.hContainer);

    return this;
  };

  /**
   * Remove the guide elements from the DOM.
   * @property _removeLines <private> [Function]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype._removeLines = function _removeLines() {
    while (this._ui.hContainer.firstChild) {
      this._ui.hContainer.removeChild(this._ui.hContainer.firstChild);
    }

    while (this._ui.vContainer.firstChild) {
      this._ui.vContainer.removeChild(this._ui.vContainer.firstChild);
    }

    return this;
  };

  /**
   * Crate a new Element (line) and append it to the parentElement passed as
   * the first argument. It will also add the styles passed as second param.
   * @property _appendLine <private> [Function]
   * @argument parent <required> [DOMElement] the element to append the line.
   * @argument styles <optional> [Object] the styles to be added to the line.
   * @return this [PSDGuides]
   */
  PSDGuides.prototype._appendLine = function _appendLine(parent, styles) {
    var line = document.createElement("div");

    this._setStyles(line, styles);
    parent.appendChild(line);

    return this;
  };

  /**
   * Return the max available width or height of the document.
   * @property _getMaxSize <private> [Function]
   * @argument prop <required> [String] "Width|Height"
   * @return Math.max(...) [Number]
   */
  PSDGuides.prototype._getMaxSize = function _getMaxSize(prop) {
    var scroll, offset, client;

    scroll = Math.max(this._ui.body['scroll' + prop], this._ui.documentElement['scroll' + prop]);
    offset = Math.max(this._ui.body['offset' + prop], this._ui.documentElement['offset' + prop]);
    client = Math.max(this._ui.body['client' + prop], this._ui.documentElement['client' + prop]);

    return Math.max(scroll, offset, client);
  };

  /**
   * Return the number of pixels for psd-guides container to be aligned.
   * @property _getAligment <private> [Function]
   * @return [Number]
   */
  PSDGuides.prototype._getAligment = function _getAligment(siteWidth) {
    if (this.config.alignment === "left") {
      return 0;
    }

    if (this.config.alignment === "right") {
      return Math.floor(this._getMaxSize("Width") - siteWidth);
    }

    return Math.floor((this._getMaxSize("Width") - siteWidth) / 2);
  };

  /**
   * Utility method for adding styles to elements.
   * @property _setStyles <private> [Function]
   * @argument element <required> [DOMElement]
   * @argument hash <required> [Object]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype._setStyles = function _setStyles(element, hash) {
    Object.keys(hash).forEach(function(propertyName) {
      element.style[propertyName] = hash[propertyName];
    });

    return this;
  };

  /**
   * CSS override utility.
   * @property _overrideCSS <private> [Function]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype._overrideCSS = function _overrideCSS() {
    var css, head, style;

    css = ".psd-guides-wrapper * {-webkit-box-sizing: border-box !important; box-sizing: border-box !important;}";
    head = document.getElementsByTagName("head")[0];
    style = document.createElement("style");

    style.type = "text/css";

    if (style.styleSheet) {
      style.styleSheet.cssText = css;
    } else {
      style.appendChild(document.createTextNode(css));
    }

    head.appendChild(style);

    return this;
  };

  PSDGuides.prototype._hide = function _hide() {
    this._ui.wrapper.style.display = "none";
  };

  /**
   * Return the current saved horizontal guides.
   * @property getHorizontalGuides <public> [Function]
   * @return this._horizontalGuides [Array]
   */
  PSDGuides.prototype.getHorizontalGuides = function getHorizontalGuides() {
    return this._horizontalGuides;
  };

  /**
   * Return the current saved vertical guides.
   * @property getVerticalGuides <public> [Function]
   * @return this._verticalGuides [Array]
   */
  PSDGuides.prototype.getVerticalGuides = function getVerticalGuides() {
    return this._verticalGuides;
  };

  /**
   * Clear the horizontal guides array reference.
   * @property removeHorizontalGuides <public> [Function]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype.removeHorizontalGuides = function removeHorizontalGuides() {
    this._horizontalGuides = [];

    return this;
  };

  /**
   * Clear the vertical guides array reference.
   * @property removeVerticalGuides <public> [Function]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype.removeVerticalGuides = function removeVerticalGuides() {
    this._verticalGuides = [];

    return this;
  };

  /**
   * Clear both horizontal and vertical array references.
   * @property removeGuides <public> [Function]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype.removeGuides = function removeGuides() {
    this.removeHorizontalGuides().removeVerticalGuides();

    return this;
  };

  /**
   * Update the width and height of psd-guides container,
   * remove and create the guides using the guides array references.
   * @property update <public> [Function]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype.update = function update() {
    this._setStyles(this._ui.wrapper, {
      width: this._getMaxSize("Width") + "px",
      height: this._getMaxSize("Height") + "px"
    })._removeLines()._createVerticalLines()._createHorizontalLines();

    return this;
  };

  /**
   * Display the guides.
   * @property activate <public> [Function]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype.activate = function activate() {
    this.active = true;
    this.update();
    this._ui.wrapper.style.display = "";

    return this;
  };

  /**
   * Hide the guides.
   * @property deactivate <public> [Function]
   * @return this [PSDGuides]
   */
  PSDGuides.prototype.deactivate = function deactivate() {
    this.active = false;
    this._hide();

    return this;
  };

  /**
   * Clean all references to other objects and remove DOMElements.
   * @property destroy <public> [Function]
   * @return null
   */
  PSDGuides.prototype.destroy = function destroy () {
    this._removeLines();
    this._ui.wrapper.removeChild(this._ui.vContainer);
    this._ui.wrapper.removeChild(this._ui.hContainer);
    this.config.canvas.removeChild(this._ui.wrapper);
    window.removeEventListener("resize", this.__resizeHandler, false);

    Object.keys(this).forEach(function(propertyName) {
      delete this[propertyName];
    }, this);

    return null;
  };

  if (typeof exports === 'object') {module.exports = PSDGuides;}
  else {window.PSDGuides = PSDGuides;}
})();

},{}],2:[function(require,module,exports){
(function() {
  var PSDGuides = require('psd-guides')
    , uiControl = [].slice.call(document.querySelectorAll('.psdguides-ctrl'), 0)
    , fixedControl = document.querySelector('.psdguides-ctrl--fixed')
    , guides = new PSDGuides({
      canvasWidth: 500,
      lineColor: "#0fc",
      horizontalGuides: ['(10, 60) * 7'],
      verticalGuides: [37, 70, 25, 55, 20, 50, 35, 3, 40, 15, 30, 50, 55, 130, 90, 110, '(55, 110) * 2', 35, 3, 155]
    });

  uiControl.forEach(function(ctrl) {
    ctrl.addEventListener('click', toggle, false);
  });

  function toggle(ev) {
    ev.preventDefault();

    if (guides.active) {
      fixedControl.textContent = 'Show PSDGuides.js';
      return guides.deactivate();
    }

    fixedControl.textContent = 'Hide PSDGuides.js';
    return guides.activate();
  }
})();

},{"psd-guides":1}]},{},[2]);
